src/overlays/actors/ovl_En_Rat/z_en_rat.c

```diff
@@ -28,123 +47,913 @@ const ActorInit En_Rat_InitVars = {
+                posB.y = posA.y - (this->axisLeft.y * lineLength);
+                posB.z = posA.z - (this->axisLeft.z * lineLength);
+            }
+
+            if (EnRat_IsOnCollisionPoly(play, &posA, &posB, &posSide, &polySide, &bgIdSide)) {
+                this->shouldRotateOntoSurfaces |= EnRat_UpdateFloorPoly(this, polySide, play);
+                Math_Vec3f_Copy(&this->actor.world.pos, &posSide);
+                this->actor.floorBgId = bgIdSide;
+                break;
+            }
+        }
+
+        if (i == 3) {
+            // no collision nearby
+            return false;
+        }
+    }
+
+    return true;
+}
+
+/**
+ * Transform coordinates from actor coordinate space to world space, according to current orientation.
+ * `offset` is expected to already be at world scale.
+ */
+void EnRat_ActorCoordsToWorld(EnRat* this, Vec3f* offset, Vec3f* pos) {
+    f32 x = offset->x + this->visualJitter;
+
+    pos->x = this->actor.world.pos.x + (this->axisLeft.x * x) + (this->axisUp.x * offset->y) +
+             (this->axisForwards.x * offset->z);
+    pos->y = this->actor.world.pos.y + (this->axisLeft.y * x) + (this->axisUp.y * offset->y) +
+             (this->axisForwards.y * offset->z);
+    pos->z = this->actor.world.pos.z + (this->axisLeft.z * x) + (this->axisUp.z * offset->y) +
+             (this->axisForwards.z * offset->z);
+}
+
+/**
+ * This function will spawn splashes and ripples as the Real Bombchu runs across the water's surface.
+ */
+void EnRat_SpawnWaterEffects(EnRat* this, PlayState* play) {
+    s32 pad;
+    Vec3f splashPos;
+
+    EffectSsGRipple_Spawn(play, &this->actor.world.pos, 70, 500, 0);
+    splashPos.x = this->actor.world.pos.x - (this->axisForwards.x * 10.0f);
+    splashPos.z = this->actor.world.pos.z - (this->axisForwards.z * 10.0f);
+    splashPos.y = this->actor.world.pos.y + 5.0f;
+    EffectSsGSplash_Spawn(play, &splashPos, NULL, NULL, 1, 450);
+}
+
+void EnRat_HandleNonSceneCollision(EnRat* this, PlayState* play) {
+    s16 yaw = this->actor.shape.rot.y;
+    f32 sin;
+    f32 cos;
+    f32 tempX;
+
+    BgCheck2_UpdateActorAttachedToMesh(&play->colCtx, this->actor.floorBgId, &this->actor);
+
+    if (yaw != this->actor.shape.rot.y) {
+        yaw = this->actor.shape.rot.y - yaw;
+
+        sin = Math_SinS(yaw);
+        cos = Math_CosS(yaw);
+
+        tempX = this->axisForwards.x;
+        this->axisForwards.x = (sin * this->axisForwards.z) + (cos * tempX);
+        this->axisForwards.z = (cos * this->axisForwards.z) - (sin * tempX);
+
+        tempX = this->axisUp.x;
+        this->axisUp.x = (sin * this->axisUp.z) + (cos * tempX);
+        this->axisUp.z = (cos * this->axisUp.z) - (sin * tempX);
+
+        tempX = this->axisLeft.x;
+        this->axisLeft.x = (sin * this->axisLeft.z) + (cos * tempX);
+        this->axisLeft.z = (cos * this->axisLeft.z) - (sin * tempX);
+    }
+}
+
+Vec3f sSmokeAccel = { 0.0f, 0.6f, 0.0f };
+Color_RGBA8 sSmokeColor = { 255, 255, 255, 255 };
```

Comment:
```suggestion
static Vec3f sSmokeAccel = { 0.0f, 0.6f, 0.0f };
static Color_RGBA8 sSmokeColor = { 255, 255, 255, 255 };
```
I feel these are generic enough that they might clash with other names. I would assume many actors could spawn smoke